### DISCUSSION TITLE
Cypress order of execution of before(), and It() blocks.

### DISCUSSION THREAD

**Discussion by Pverma96** [cypress-io/cypress#27646] [Questions and Help]
I have the below test case:

`describe('My Test Suite', () => {
  before(() => {
    cy.reportLog ("1st before")
  });

  before(() => {
     cy.reportLog ("2st before")
  });

  it('Test Case 1', () => {
   cy.reportLog ("1st it")
  });

  before(() => {
    cy.reportLog ("3rd before ")
  });

  it('Test Case 2', () => {
   cy.reportLog ("2nd it")
  });

   after(() => {
    cy.reportLog ("after")
   });

});`
I am expecting the output be 

1st before
2st before
1st it
3rd before
2nd it
after

Instead cypress Version 12.17.4 gives the output 

1st before
2st before
3rd before
1st it
2nd it
after

Is the output from cypress is expected behaviour?

**Comment by portyu9** [comment]
Hi @Pverma96 

Yes, the output you are seeing is expected.

Cypress uses Mocha's suite/hook semantics. All of the `before()` hooks registered for a `describe()` suite run **before any `it()` tests in that suite**, regardless of where the hook appears between the `it()` declarations in the source.

So this:

```js
before(...)
it(...)
before(...)
it(...)
```

does **not** mean "before, test, before, test". It registers two suite-level `before` hooks and two tests, so both `before()` hooks execute first.

If the third setup is only needed before the second test, either put the setup in that test or create a nested suite, for example:

```js
describe('My Test Suite', () => {
  before(() => {
    cy.reportLog('1st before')
  })

  it('Test Case 1', () => {
    cy.reportLog('1st it')
  })

  describe('second part', () => {
    before(() => {
      cy.reportLog('3rd before')
    })

    it('Test Case 2', () => {
      cy.reportLog('2nd it')
    })
  })
})
```

Use `beforeEach()` instead when the same setup must run before every test in a suite.

Please mark this answer as accepted if it helped, thank you.
